samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -438,11 +442,48 @@ public void publishJobModelVersion(String oldVersion, String newVersion) {
         "(actual data version after update = " + stat.getVersion() + ")");
   }
 
+  // validate Zk protocol currently used is the same as in this participant
+  private void validateZkVersion() {
+    String rootPath = keyBuilder.getRootPath();
+    if (!zkClient.exists(rootPath)) {
+      try {
+        // attempt to create the root with the correct version
+        zkClient.createPersistent(rootPath, ZK_PROTOCOL_VERSION);
+        LOG.info("Created zk root node: " + rootPath + " with zk version " + ZK_PROTOCOL_VERSION);
+        return;
+      } catch (ZkNodeExistsException e) {
+        // ignoring
+        LOG.warn("root path " + rootPath + " already exists.");
+      }
+    }
+    // if exists, verify the version
+    Stat stat = new Stat();
+    String version = (String) zkClient.readData(rootPath, stat);
+    if (version == null) {
+      // for backward compatibility, if no value - assume 1.0
+      try {
+        zkClient.writeData(rootPath, "1.0", stat.getVersion());
+      } catch (ZkBadVersionException e) {
+        // if the write failed with ZkBadVersionException it means someone else already wrote a version, so we can ignore it.
+      }
+      // re-read the updated version
+      version = (String) zkClient.readData(rootPath);
+    }
+    LOG.info("Current version for zk root node: " + rootPath + " is " + version + ", expected version is " + ZK_PROTOCOL_VERSION);
+    if (!version.equals(ZK_PROTOCOL_VERSION)) {
+      throw new SamzaException("ZK Protocol mismatch. Expected " + ZK_PROTOCOL_VERSION + "; found " + version);
+    }
+  }
+
   /**
    * verify that given paths exist in ZK
    * @param paths - paths to verify or create
    */
-  public void makeSurePersistentPathsExists(String[] paths) {
+  public void validatePaths(String[] paths) {
```

Comment:
It is minor, but to me this new name is not very accurate. I would call it assurePathExists.